parser.py

```diff
@@ -1,17 +1,30 @@
 #!/bin/env python
 import argparse
 import pprint
+import re
+
 import grammar
 import visitor
 
+class PreProcessor:
+    @staticmethod
+    def remove_comments(program: str):
+        lines = program.splitlines()
+        cleaned_lines = []
+        for line in lines:
+            cleaned_line = re.sub(r'#.*$', '', line).strip()
+            if cleaned_line:
```

Comment:
In what cases can cleaned line be false? This doesn't look good either. If I understand correctly you are removing empty lines. The same way as with `strip` you lose important information and make the parser unable to correctly report errors because line numbers will no longer match. Every line in the original should be there in the preprocessed text.

Reply:
Added this due to issues with some comments (e.g., comments before the subroutine or at the end of the text) that were causing problems with the grammar, presumably because they leave line breaks that we don't allow.